docs/enhancements/0002_day0-integration.md

```diff
@@ -0,0 +1,285 @@
+  generation: 1
+  labels:
+    machineconfiguration.openshift.io/role: worker
+  name: 99-ybettan-external-image
+  resourceVersion: "1758"
+  uid: d1f07169-4cb5-4b85-93b3-3cd80530f028
+spec:
+  osImageURL: quay.io/ybettan/rhcos:414.92.202312311229-0
+```
+The important piece in it is `osImageURL: quay.io/ybettan/rhcos:414.92.202312311229-0`
+which tells MCO what that the ostree-commit that is supposed to be running on
+the node is encapsulated in `quay.io/ybettan/rhcos:414.92.202312311229-0`.
+
+If MCO, finds out that the OS on the node is different than the image specified
+in `osImageURL` then MCO will override the node OS based on the `MachineConfig`
+and reboot - we are going to leverage this mechanisem for the integration with
+KMM.
+
+In a day-0-cluster scenario, the user has applied this `MachineConfig` to tell
+MCO he has modified the OS on the nodes and that it shoudn't override it with
+the RHCOS image from the payload.
+
+### Setting up the integration
+
+A user that wishes to hand over the management of his day-0 kmod to KMM will
+first need to make sure the `MachineConfig` can be reconize by KMM, therefore,
+he would add a `machineconfiguration.openshift.io/kmm-managed: "true"` label to
+it.
+
+KMM will watch for `MachineConfig`s in the cluster containing the
+`machineconfiguration.openshift.io/kmm-managed: "true"` label and create a new
+`ModuleDay0` CR in the cluster for each one of them.
+We will have a different controller for this new CRD in KMM.
+**There should be no more than 1 `ModuleDay0` targeting a `MachineConfigPool`
+(masters/workers).
+If multiple kmods are required they should all be backed in the same image**
+
+Since MCO can only apply `MachineConfig`s changes to an entire `MachineConfigPool`,
+we must require that a `ModuleDay0` will target a `MachineConfigPool` as well.
+MCO ties between the `machineconfiguration.openshift.io/role: worker` label on
+the `MachineConfig` and the `node-role.kubernetes.io/worker: ""` label on the
+node.
+
+A customer can always create his own `MachineConfigPool` and add only 1 worker
+to it but I don't think this should be one of our primary concerns ATM for the
+following reasons
+  * assisted-installer will generate the same ISO for all nodes in the cluster
+    anyway so there is no benefit of having finer grained node selector for
+    `ModuleDay0` when assisted doesn't support such resolution on its own.
+  * if a user wishes to have only some worker nodes modified, he will need to
+    create is own `MachineConfigPool` post cluster installation and then update
+    the `machineconfiguration.openshift.io/role: worker` label of the
+    `MachineConfig` to target the new `MachineConfigPool`.
+    This will make MCO role back the nodes that shouldn't be updated to the
+    payload OS.
+    Using layering for only some of the workers will result in a cluster upgrade
+    that only upgrade the nodes that aren't layered which may cause some
+    inconsistency and confusion
```

Comment:
Can assisted use the custom ISO for some nodes only?

Reply:
It can but not in the way that assisted was designed to.
It will make more mess than order.
Should be simpler to just install the driver on all workers even if the node doesn't have the HW.
In addition to that, if an admin wish to add the HW later on then the driver is already in place.

If a user really want to be as precise as possible he can always follow:
1. Make all worker nodes custom
2. Once the cluster is installed, the user create a `MachineConfigPool` and add the worker he wishes to customize to it.
3. He modifies the `MachineConfig` to target the new pool he has created.
4. MCO reboot all other worker nodes to the default RHCOS.

https://redhat-internal.slack.com/archives/CUPJTHQ5P/p1705236817661129

Reply:
it does not make sense to maintain iso per node. it's not scalable.

Reply:
> it does not make sense to maintain iso per node. it's not scalable.

I agree. I just mentioned that to answer a potential question that says "but what if a customer really wan't to do it" - I have removed this whole section from the proposal.